examples/tv-casting-app/linux/main.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/server/Mdns.h>
+#include <app/server/OnboardingCodesUtil.h>
+#include <app/server/Server.h>
+#include <controller/CHIPCommissionableNodeController.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/ConfigurationManager.h>
+#include <setup_payload/SetupPayload.h>
+#include <system/SystemLayer.h>
+#include <transport/raw/PeerAddress.h>
+
+using namespace chip;
+
+#define TV_DEVICE_TYPE 35
+
+chip::Controller::CommissionableNodeController commissionableNodeController;
+
+void RequestCommissioning(intptr_t commandArg)
+{
+    int commissionerCount = 0, selectedCommissionerNumber = CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES;
+    chip::SetupPayload payload;
+    const Mdns::DiscoveredNodeData * selectedCommissioner = nullptr;
+
+    // Display discovered commissioner TVs to ask user to select one
+    for (int i = 0; i < CHIP_DEVICE_CONFIG_MAX_DISCOVERED_NODES; i++)
+    {
+        const Mdns::DiscoveredNodeData * commissioner = commissionableNodeController.GetDiscoveredCommissioner(i);
+        if (commissioner != nullptr)
+        {
+            ChipLogProgress(Zcl, "Discovered Commissioner #%d", ++commissionerCount);
+            commissioner->LogDetail();
+        }
+    }
+    ChipLogProgress(Zcl, "%d commissioner(s) discovered", commissionerCount);
+
+    if (commissionerCount > 0)
+    {
+        ChipLogProgress(Zcl, "Choose a commissioner (by number# above) to request commissioning from: ");
+        scanf("%d", &selectedCommissionerNumber);
```

Comment:
Capturing some discussion that occurred over slack.

The app is now mostly restructured to operate in a single event loop.  However, this scanf capture of user input is still blocking.

One way to solve this is to have the system layer asynchronously watch stdin on behalf of the app, with the result that this method, `RequestCommissioning`, would be split in two: the first half would execute as it does here by callback from ScheduleWork.

The second half would be triggered as one or more callbacks when read of stdin unblocks, until enough user input is provided for the second half of this method to execute.  With these changes, I believe the app would be fully async.

@kpschoedel provided these pre and post #9086 recipes to do this:

```
// without #9086
System::WatchableSocket t;
t.Init(sl.WatchableEventsManager());
t.Attach(fd);
t.SetCallback(callback, data);
t.RequestCallbackOnPendingRead();
// and when done
t.ReleaseFD()
```

```
// with #9086
System::Layer as sl:
SocketWatchToken t;
sl.StartWatchingSocket(fd, &t);
slSetCallback(t, callback, data);
sl.RequestCallbackOnPendingRead(t);
sl.StopWatchingSocket(&t);
```

I am not personally insisting that this has to happen before merge.  Maybe it's arguable it could be a TODO.